File: cpp/Convert-Binary-Search-Tree-To-Sorted-Doubly-Lnked-List/Convert-Binary-Search-Tree-To-Sorted-Doubly-Lnked-List.cpp
class Solution {
public:
    Node* treeToDoublyList(Node* root) {
        if (root==NULL)
            return NULL;
        Node* pre = NULL;
        helper(root, pre);
        Node* head =pre;
        while(head && head->left)
        {
            head = head->left;
        }
        pre -> right = head;
        head->left = pre;
        return head;
    }
    void helper(Node* cur, Node* &pre)
    {
        if (cur==NULL)
            return;
        helper(cur->left, pre);
        cur->left = pre;
        if (pre!=NULL)
            pre->right = cur;
        pre = cur;
        helper(cur->right, pre);
    }
};

File: python/Convert-Binary-Search-Tree-To-Sorted-Doubly-Lnked-List/Convert-Binary-Search-Tree-To-Sorted-Doubly-Lnked-List.py
class Solution:
    def treeToDoublyList(self, root: 'Node') -> 'Node':
        if root  is None:
            return None
        pre = None
        pre = self.helper(root, pre)
        head = pre
        while head and head.left:
            head = head.left
        pre.right = head
        head.left = pre
        return head
    def helper(self, cur,  pre):
        if cur == None:
            return pre
        pre = self.helper(cur.left, pre)
        cur.left = pre
        if pre:
            pre.right = cur
        pre = cur
        return self.helper(cur.right, pre)
